types/scene/surface_tree.c

```diff
@@ -0,0 +1,206 @@
+			current.link) {
+		struct wlr_scene_surface_tree *child =
+			surface_tree_from_subsurface(surface_tree, subsurface);
+		wlr_scene_node_place_above(&child->tree->node, prev);
+		prev = &child->tree->node;
+
+		wlr_scene_node_set_position(&child->tree->node,
+			subsurface->current.x, subsurface->current.y);
+	}
+}
+
+static void surface_tree_handle_surface_commit(struct wl_listener *listener,
+		void *data) {
+	struct wlr_scene_surface_tree *surface_tree =
+		wl_container_of(listener, surface_tree, surface_commit);
+
+	// TODO: only do this on subsurface order or position change
+	surface_tree_reconfigure(surface_tree);
+}
+
+static void surface_tree_addon_destroy(struct wlr_addon *addon) {
+	struct wlr_scene_surface_tree *surface_tree =
+		wl_container_of(addon, surface_tree, surface_addon);
+	wlr_scene_node_destroy(&surface_tree->tree->node);
+}
+
+static const struct wlr_addon_interface surface_tree_addon = {
+	.name = "wlr_scene_surface_tree",
+	.destroy = surface_tree_addon_destroy,
+};
+
+static struct wlr_scene_surface_tree *scene_surface_tree_create(
+	struct wlr_scene_node *parent, struct wlr_surface *surface);
+
+static void surface_tree_create_subsurface(
+		struct wlr_scene_surface_tree *surface_tree,
+		struct wlr_subsurface *subsurface) {
+	struct wlr_scene_surface_tree *subsurface_tree = scene_surface_tree_create(
+		&surface_tree->tree->node, subsurface->surface);
+	if (surface_tree == NULL) {
+		return;
+	}
+
+	subsurface_tree->parent = surface_tree;
+	wlr_addon_init(&subsurface_tree->surface_addon,
+		&subsurface->surface->addons, surface_tree, &surface_tree_addon);
+}
+
+static void surface_tree_handle_surface_new_subsurface(
+		struct wl_listener *listener, void *data) {
+	struct wlr_scene_surface_tree *surface_tree =
+		wl_container_of(listener, surface_tree, surface_new_subsurface);
+	struct wlr_subsurface *subsurface = data;
+	surface_tree_create_subsurface(surface_tree, subsurface);
+}
+
+static struct wlr_scene_surface_tree *scene_surface_tree_create(
+		struct wlr_scene_node *parent, struct wlr_surface *surface) {
+	struct wlr_scene_surface_tree *surface_tree = calloc(1, sizeof(*surface_tree));
+	if (surface_tree == NULL) {
+		return NULL;
+	}
+
+	surface_tree->tree = wlr_scene_tree_create(parent);
+	if (surface_tree->tree == NULL) {
+		free(surface_tree);
+		return NULL;
+	}
+
+	surface_tree->scene_surface =
+		wlr_scene_surface_create(&surface_tree->tree->node, surface);
+	if (surface_tree->scene_surface == NULL) {
+		wlr_scene_node_destroy(&surface_tree->tree->node);
+		free(surface_tree);
+		return NULL;
+	}
+
+	surface_tree->surface = surface;
+
+	struct wlr_subsurface *subsurface;
+	wl_list_for_each(subsurface, &surface->current.subsurfaces_below,
+			current.link) {
+		surface_tree_create_subsurface(surface_tree, subsurface);
+	}
+	wl_list_for_each(subsurface, &surface->current.subsurfaces_above,
+			current.link) {
+		surface_tree_create_subsurface(surface_tree, subsurface);
+	}
```

Comment:
Hmm, bubbling up the OOM here would be a massive pain, especially without leaking memory, but silently ignoring subtrees if allocation fails doesn't feel good either.

Reply:
Hm yeah it's a bit annoying, especially since we can't easily iterate over the stuff we allocate. Pushed a commit to bubble up the OOM without cleaning up everything.

Reply:
Hmm, I'm looking at this more closely and I think your current code does clean up everything correctly in fact. The `wlr_scene_node_destroy()` call will trigger any successfully allocated subtrees to be recursively destroyed.